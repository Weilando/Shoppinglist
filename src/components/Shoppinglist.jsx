import React from 'react';
import PropTypes from 'prop-types';
import { connect } from "react-redux";
import { Entry, EntryMode, EntryStatus } from './Entry';
import NewEntry from './NewEntry';
import InfoBox from './InfoBox';

class Shoppinglist extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      mode: EntryMode.DISPLAY,
    };

    this.toggleMode = this.toggleMode.bind(this);
    this.countDone = this.countDone.bind(this);
  }

  static propTypes = {
    entries: PropTypes.array,
  }

  static defaultProps = {
    entries: [],
  }

  toggleMode() {
    switch (this.state.mode) {
      case EntryMode.DISPLAY:
        this.setState({mode: EntryMode.EDIT});
        break;
      case EntryMode.EDIT:
      default:
        this.setState({mode: EntryMode.DISPLAY});
    }
  }

  countDone() {
    const doneEntries = this.props.entries.filter(entry => entry.status === EntryStatus.DONE);
    console.log(doneEntries.length);
    return Number(doneEntries.length);
  }

  render() {
    const entries = this.props.entries;
    const keys = [...Array(entries.length).keys()]; // Array with keys from 0 to entries.length

    let entryList = keys.map((tmpKey) =>
      <Entry
        key = {tmpKey}
        eId = {Number(entries[tmpKey].id)}
        eContent = {String(entries[tmpKey].content)}
        status = {String(entries[tmpKey].status)}
        mode = {this.state.mode}
      />
    );

    return (
      <div className="Shoppinglist">
        <form>
          <label htmlFor="modeButton">Switch to mode: </label>
          <input
            type="button"
            id="modeButton"
            className="entry"
            value={this.state.mode === EntryMode.DISPLAY ? "Edit" : "Display"}
            onClick={this.toggleMode}
          />
        </form>

        <ul className="Shoppinglist">
          {entryList}

          <NewEntry
            key="-1"
          />
        </ul>

        <InfoBox
          entryCount={this.props.entries.length}
          doneEntryCount={this.countDone()}
        />
      </div>
    );
  }
}

const mapStateToProps = state => {
  return {
    entries: state.general.entryList,
  }
}

export default connect(mapStateToProps)(Shoppinglist);
