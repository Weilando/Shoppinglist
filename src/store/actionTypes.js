export const ADD_ENTRY = 'ADD_ENTRY';
export const UPDATE_ENTRY = 'UPDATE_ENTRY';
export const DELETE_ENTRY = 'DELETE_ENTRY';
export const TOGGLE_STATUS = 'TOGGLE_STATUS';

export const UPDATE_MEAL_SUGGESTIONS = 'UPDATE_MEAL_SUGGESTIONS';
export const REMOVE_DUPLICATES = 'REMOVE_DUPLICATES';
