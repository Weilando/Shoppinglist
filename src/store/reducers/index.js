import { generalReducer } from './generalReducer';
import { mealSuggestionsReducer } from './mealSuggestionsReducer';
import { combineReducers } from 'redux';

// rootReducer
export default combineReducers({
    general: generalReducer,
    mealSuggestions: mealSuggestionsReducer
});
