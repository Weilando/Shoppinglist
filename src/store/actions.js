import { ADD_ENTRY, UPDATE_ENTRY, DELETE_ENTRY, TOGGLE_STATUS, UPDATE_MEAL_SUGGESTIONS, REMOVE_DUPLICATES } from './actionTypes';

let nextEntryId = 0;

export function addEntry(newContent) {
  return {
    type: ADD_ENTRY,
    payload: {
      id: ++nextEntryId,
      newContent
    }
  }
}

export function updateEntry(id, newContent) {
  return {
    type: UPDATE_ENTRY,
    payload: {
      id: id,
      newContent
    }
  }
}

export function deleteEntry(id) {
  return {
    type: DELETE_ENTRY,
    payload: { id: id }
  }
}

export function toggleStatus(id) {
  return {
    type: TOGGLE_STATUS,
    payload: { id: id }
  }
}

export function updateMealSuggestions(mealSuggestions) {
  return {
    type: UPDATE_MEAL_SUGGESTIONS,
    payload: { newMealSuggestions: mealSuggestions }
  }
}

export function removeDuplicates() {
  return {
    type: REMOVE_DUPLICATES
  }
}
