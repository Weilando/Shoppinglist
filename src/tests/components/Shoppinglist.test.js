import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import store from '../../store/store';
import Shoppinglist from '../../components/Shoppinglist';

describe('snapshot tests for Shoppinglist component', () => {
  it('should render empty Shoppinglist component correctly', () => {
    const tree = renderer.create(
      <Provider store={store}>
        <Shoppinglist />
      </Provider>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
