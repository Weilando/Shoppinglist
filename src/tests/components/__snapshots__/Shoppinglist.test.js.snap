// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot tests for Shoppinglist component should render empty Shoppinglist component correctly 1`] = `
<div
  className="Shoppinglist"
>
  <form>
    <label
      htmlFor="modeButton"
    >
      Switch to mode: 
    </label>
    <input
      className="entry"
      id="modeButton"
      onClick={[Function]}
      type="button"
      value="Edit"
    />
  </form>
  <ul
    className="Shoppinglist"
  >
    <li>
      <form
        onSubmit={[Function]}
      >
        <input
          className="entry"
          onChange={[Function]}
          type="text"
          value=""
        />
        <input
          className="entry"
          disabled={true}
          type="submit"
          value="Add"
        />
      </form>
    </li>
  </ul>
  <div
    className="InfoBox"
  >
    <p>
      Entries: 
      0
       â€ƒ Done entries: 
      0
    </p>
  </div>
  <div
    className="InfoBox"
  >
    <h3>
      Meal suggestions:
    </h3>
    <ul
      className="Shoppinglist"
    />
  </div>
</div>
`;
